geode-core/src/distributedTest/java/org/apache/geode/cache/query/dunit/PdxMultiThreadQueryDUnitTest.java

```diff
@@ -0,0 +1,271 @@
+            logger.info("### Executing Query on server: " + queryString[1]);
+            Query query = remoteQueryService.newQuery(queryString[1]);
+            SelectResults rs2 = (SelectResults) query.execute();
+            assertEquals(numberOfEntries, rs2.size());
+          });
+    }
+
+    for (int i = 0; i < size; i++) {
+      asyncInvocationArray[i].await();
+    }
+    vm3.invoke(() -> {
+      await().until(() -> TestObject.numInstance.get() == size * numberOfEntries);
+    });
+
+    this.closeClient(vm3);
+    this.closeClient(vm2);
+    this.closeClient(vm1);
+    this.closeClient(vm0);
+  }
+
+  @Test
+  public void testRetryWithPdxSerializationException()
+      throws CacheException, InterruptedException {
+    final Host host = Host.getHost(0);
+    final String hostName = NetworkUtils.getServerHostName();
+
+    VM vm0 = host.getVM(VersionManager.CURRENT_VERSION, 0);
+    VM vm1 = host.getVM(VersionManager.CURRENT_VERSION, 1);
+    VM vm2 = host.getVM(VersionManager.CURRENT_VERSION, 2);
+    VM vm3 = host.getVM(VersionManager.CURRENT_VERSION, 3);
+
+    // Start servers
+    for (VM vm : Arrays.asList(vm0, vm1, vm2)) {
+      vm.invoke(() -> {
+        configAndStartBridgeServer();
+      });
+    }
+
+    // create pdx instance at servers
+    vm0.invoke(() -> {
+      Region region = getRootRegion().getSubregion(regionName);
+      for (int i = 0; i < numberOfEntries; i++) {
+        region.put("key-" + i, new TestPdxSerializable());
+      }
+    });
+
+    final int port0 = vm0.invoke(() -> PdxQueryDUnitTest.getCacheServerPort());
+    final int port1 = vm1.invoke(() -> PdxQueryDUnitTest.getCacheServerPort());
+    final int port2 = vm2.invoke(() -> PdxQueryDUnitTest.getCacheServerPort());
+
+    // Create client pool.
+    final String poolName = "testClientServerQueryPool";
+    createPool(vm3, poolName, new String[] {hostName, hostName, hostName},
+        new int[] {port0, port1, port2}, true);
+
+    try {
+      vm3.invoke(() -> {
+        TestPdxSerializable.throwExceptionOnDeserialization = true;
+        QueryService remoteQueryService = (PoolManager.find(poolName)).getQueryService();
+        logger.info("### Executing Query on server: " + queryString[1]);
+        Query query = remoteQueryService.newQuery(queryString[1]);
+        SelectResults rs2 = (SelectResults) query.execute();
+        assertEquals(numberOfEntries, rs2.size());
+      });
+    } finally {
+      vm3.invoke(() -> {
+        assertEquals(false, TestPdxSerializable.throwExceptionOnDeserialization);
+        // the failed try will increment numInstance once
+        assertEquals(numberOfEntries + 1, TestPdxSerializable.numInstance.get());
+        TestPdxSerializable.numInstance.set(0);
+      });
+    }
+
+    this.closeClient(vm3);
+    this.closeClient(vm2);
+    this.closeClient(vm1);
+    this.closeClient(vm0);
+  }
+
+  public static class TestPdxSerializable implements PdxSerializable {
```

Comment:
The fact that `TestObject` and `TestObject2` are `PdxSerializable` makes me confused as to why this new object is called `TestPdxSerializable`.

I recommend a more descriptive class name like `TestObjectThrowsPdxSerializationException`.